File: srcs/server/db/rls_e2e_test.go
package db

import (
	"context"
	"database/sql"
	"os"
	"testing"
	"time"

	_ "github.com/lib/pq"
	"github.com/stretchr/testify/assert"
)

func TestPostgresRLSE2EIntegration(t *testing.T) {
	dbURL := os.Getenv("DATABASE_URL")
	if dbURL == "" {
		dbURL = "postgres://postgres:postgres@localhost:5432/test?sslmode=disable"
	}

	db, err := sql.Open("postgres", dbURL)
	if err != nil {
		t.Skipf("Skipping integration test: %v", err)
	}
	defer db.Close()

	if err := db.Ping(); err != nil {
		t.Skipf("Skipping integration test due to ping failure: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// 1. Setup tables
	setupSQL := `
		DROP TABLE IF EXISTS rls_e2e_agents;
		CREATE TABLE rls_e2e_agents (
			id TEXT PRIMARY KEY,
			tenant_id TEXT NOT NULL,
			department TEXT NOT NULL
		);
		ALTER TABLE rls_e2e_agents ENABLE ROW LEVEL SECURITY;
		ALTER TABLE rls_e2e_agents FORCE ROW LEVEL SECURITY;
		DROP POLICY IF EXISTS rls_e2e_agents_policy ON rls_e2e_agents;
		CREATE POLICY rls_e2e_agents_policy ON rls_e2e_agents
			FOR ALL
			USING (tenant_id = nullif(current_setting('app.current_tenant', true), '')::text);
	`
	_, err = db.ExecContext(ctx, setupSQL)
	assert.NoError(t, err)

	// 2. Insert test data
	disableSQL := `ALTER TABLE rls_e2e_agents NO FORCE ROW LEVEL SECURITY;`
	_, err = db.ExecContext(ctx, disableSQL)
	assert.NoError(t, err)

	insertSQL := `
		INSERT INTO rls_e2e_agents (id, tenant_id, department) VALUES
		('a1', 'tenant_A', 'sales'),
		('a2', 'tenant_B', 'support')
	`
	_, err = db.ExecContext(ctx, insertSQL)
	assert.NoError(t, err)

	enableSQL := `ALTER TABLE rls_e2e_agents FORCE ROW LEVEL SECURITY;`
	_, err = db.ExecContext(ctx, enableSQL)
	assert.NoError(t, err)

	setupRoleSQL := `
		DO $$
		BEGIN
		  IF NOT EXISTS (
			SELECT FROM pg_catalog.pg_roles
			WHERE  rolname = 'rls_test_user') THEN
			CREATE ROLE rls_test_user LOGIN;
		  END IF;
		END
		$$;
		GRANT ALL PRIVILEGES ON TABLE rls_e2e_agents TO rls_test_user;
	`
	_, err = db.ExecContext(ctx, setupRoleSQL)
	assert.NoError(t, err)

	// 3. Test Tenant A
	connA, err := db.Conn(ctx)
	assert.NoError(t, err)
	defer connA.Close()

	_, err = connA.ExecContext(ctx, "SET ROLE rls_test_user")
	assert.NoError(t, err)
	_, err = connA.ExecContext(ctx, "SELECT set_config('app.current_tenant', 'tenant_A', true)")
	assert.NoError(t, err)

	rowsA, err := connA.QueryContext(ctx, "SELECT id, department FROM rls_e2e_agents")
	assert.NoError(t, err)
	defer rowsA.Close()

	var countA int
	for rowsA.Next() {
		countA++
	}
	assert.Equal(t, 1, countA, "Tenant A should see exactly 1 agent")

	// 4. Test API manipulation (Tenant A trying to read without setting context / bad context)
	connManipulate, err := db.Conn(ctx)
	assert.NoError(t, err)
	defer connManipulate.Close()

	_, err = connManipulate.ExecContext(ctx, "SET ROLE rls_test_user")
	assert.NoError(t, err)
	_, err = connManipulate.ExecContext(ctx, "SELECT set_config('app.current_tenant', '', true)")
	assert.NoError(t, err)

	rowsManipulate, err := connManipulate.QueryContext(ctx, "SELECT id FROM rls_e2e_agents")
	assert.NoError(t, err)
	defer rowsManipulate.Close()

	var countEmpty int
	for rowsManipulate.Next() {
		countEmpty++
	}
	assert.Equal(t, 0, countEmpty, "Empty tenant context should see 0 agents")

	// Cleanup
	_, err = db.ExecContext(ctx, "DROP TABLE IF EXISTS rls_e2e_agents")
	assert.NoError(t, err)
}

File: src/server/rls_e2e_test.rs
use crate::db::{init_db, DbStore, get_pg_pool};
use std::env;
use serial_test::serial;

#[tokio::test]
#[serial]
async fn test_postgres_rls_e2e_isolation() {
    let mode = env::var("OHC_STANDALONE").unwrap_or_default();
    if mode == "true" {
        return; // Skip in SQLite standalone mode
    }

    env::set_var("DATABASE_URL", "postgres://postgres:postgres@localhost:5432/test?sslmode=disable");
    let store = init_db().await.unwrap();

    if let DbStore::Postgres = store {
        let pool = get_pg_pool().await;

        let tenant_a = uuid::Uuid::new_v4();
        let tenant_b = uuid::Uuid::new_v4();

        // Ensure RLS table exists via init_db above

        // Insert data bypassing RLS
        let mut tx = pool.begin().await.unwrap();
        sqlx::query("ALTER TABLE agents DISABLE ROW LEVEL SECURITY;").execute(&mut *tx).await.unwrap();
        sqlx::query("INSERT INTO agents (id, tenant_id, department, status) VALUES ($1, $2, 'sales', 'active')")
            .bind(uuid::Uuid::new_v4())
            .bind(tenant_a)
            .execute(&mut *tx).await.unwrap();

        sqlx::query("INSERT INTO agents (id, tenant_id, department, status) VALUES ($1, $2, 'support', 'active')")
            .bind(uuid::Uuid::new_v4())
            .bind(tenant_b)
            .execute(&mut *tx).await.unwrap();
        sqlx::query("ALTER TABLE agents ENABLE ROW LEVEL SECURITY;").execute(&mut *tx).await.unwrap();
        tx.commit().await.unwrap();

        // 1. Query as Tenant A
        let mut tx_a = pool.begin().await.unwrap();
        sqlx::query(&format!("SELECT set_config('app.current_tenant', '{}', true)", tenant_a)).execute(&mut *tx_a).await.unwrap();
        let count_a: (i64,) = sqlx::query_as("SELECT count(*) FROM agents").fetch_one(&mut *tx_a).await.unwrap();
        assert_eq!(count_a.0, 1, "Tenant A should see exactly 1 agent");

        // 2. Query as Tenant B
        let mut tx_b = pool.begin().await.unwrap();
        sqlx::query(&format!("SELECT set_config('app.current_tenant', '{}', true)", tenant_b)).execute(&mut *tx_b).await.unwrap();
        let count_b: (i64,) = sqlx::query_as("SELECT count(*) FROM agents").fetch_one(&mut *tx_b).await.unwrap();
        assert_eq!(count_b.0, 1, "Tenant B should see exactly 1 agent");

        // 3. Query without Tenant Context
        let mut tx_empty = pool.begin().await.unwrap();
        sqlx::query("SELECT set_config('app.current_tenant', '', true)").execute(&mut *tx_empty).await.unwrap();
        let count_empty: (i64,) = sqlx::query_as("SELECT count(*) FROM agents").fetch_one(&mut *tx_empty).await.unwrap();
        assert_eq!(count_empty.0, 0, "Empty context should see 0 agents");
    }
}
